packages/mongodb-log-writer/src/mongo-log-manager.ts

```diff
@@ -69,23 +77,58 @@ export class MongoLogManager {
       if (!id) continue;
       const fileTimestamp = +new ObjectId(id).getTimestamp();
       const fullPath = path.join(dir, dirent.name);
-      let toDelete: string | undefined;
+      let toDelete:
+        | {
+            fullPath: string;
+            /** If the file wasn't deleted right away and there is a
+             *  retention size limit, its size should be accounted */
+            fileSize?: number;
+          }
+        | undefined;
 
       // If the file is older than expected, delete it. If the file is recent,
       // add it to the list of seen files, and if that list is too large, remove
       // the least recent file we've seen so far.
       if (fileTimestamp < deletionCutoffTimestamp) {
-        toDelete = fullPath;
-      } else if (this._options.maxLogFileCount) {
-        leastRecentFileHeap.push({ fullPath, fileTimestamp });
-        if (leastRecentFileHeap.size() > this._options.maxLogFileCount) {
-          toDelete = leastRecentFileHeap.pop()?.fullPath;
+        toDelete = {
+          fullPath,
+        };
+      } else if (this._options.retentionGB || this._options.maxLogFileCount) {
+        let fileSize: number | undefined;
+
+        if (this._options.retentionGB) {
+          try {
+            fileSize = (await fs.stat(fullPath)).size;
+            if (this._options.retentionGB) {
+              usedStorageSize += fileSize;
+            }
+          } catch (err) {
+            this._options.onerror(err as Error, fullPath);
+          }
+        }
+
+        leastRecentFileHeap.push({
+          fullPath,
+          fileTimestamp,
+          fileSize,
+        });
+
+        const reachedMaxStorageSize = usedStorageSize > storageSizeLimit;
+        const reachedMaxFileCount =
+          this._options.maxLogFileCount &&
+          leastRecentFileHeap.size() > this._options.maxLogFileCount;
+
+        if (reachedMaxStorageSize || reachedMaxFileCount) {
+          toDelete = leastRecentFileHeap.pop();
```

Comment:
So ... took some time to think this through, but the problem with this approach is that we'll generally want to delete the _oldest_ files that are necessary to ensure that max file count + max file size limitations

But with this approach, that's no longer the case – you could end up with e.g. the most recent file being deleted, if it happens to be the first file we read from the directory and it's already over the size limit – right?

(I'm not sure if we need to let go of the heap as the underlying data structure, certainly wouldn't be the end of the world, but I'm pretty sure that the "read one file, check conditions, maybe delete one file" approach won't work anymore)

Reply:
hmm yeah, we could sort by creation date (which would be same as the log's name?) from the getgo and adjust the logic accordingly

Reply:
> which would be same as the log's name?

Yes, that's an assumption I'd still keep around 👍